README.md
# Pianoputer

This library lets you to play your computer keyboard like a piano. Here is a [video](https://www.youtube.com/watch?v=z410eauCnHc) of it in action on a French azerty keyboard.

## Play!

Pianoputer only works in python3 so make sure you are using python3
```
pip install pianoputer
pianoputer
```

After a few seconds, the below image will appear in a window, indicating that the program is ready.
The cyan key is the key that the sample wav file is assigned to. By default this is c4, [the piano middle C at 261.6 hz](https://en.wikipedia.org/wiki/Piano_key_frequencies)
All white and black keys are transposed up and down from the anchor cyan key.

![qwerty keyboard layout, c4 is cyan](https://raw.githubusercontent.com/Zulko/pianoputer/master/pianoputer/keyboards/qwerty_piano.jpg "qwerty keyboard layout, c4 is cyan")

## Changing the sound file

You can provide your own sound file with

```
pianoputer --wav my_sound_file.wav
```
For example:
```
pianoputer -w audio_files/bowl_c6.wav
```
All white and black keys are transposed up and down from the anchor cyan key.

## Changing the keyboard layout

Note that the default keyboard configuration (stored in file `keyboards/qwerty_piano.txt`) is for the most commonly used QWERTY keyboards. You can change the configuration so that it matches your keyboard, for instance using the alternative `keyboards/azerty_typewriter.txt`:

```
pianoputer -k keyboards/azerty_typewriter.txt
```
![azerty typewriter](https://raw.githubusercontent.com/Zulko/pianoputer/master/pianoputer/keyboards/azerty_typewriter.jpg "azerty typewriter keyboard layout")
These `.txt` files simply contain a sequence of key names and are easy to edit. For convenience this repository also provides a `make_kb_file.py` program:
```
python make_kb_file.py
```

This will let you press the keys in the order that you want, and create a new keyboard configuration file, by default `my_keyboard.kb` (just follow the instructions). You can then use the custom keyboard file with the --keyboard argument

## Local Installation
```
python3 -m venv venv
source venv/bin/activate
# if you want to edit the program and have pianoputer use your edits
pip install -e .
# to install pianoputer separately in your virtual environment
pip install .
pianoputer
```

## Attributions
- c4 piano sample from https://en.wikipedia.org/wiki/File:Middle_C.mid

## Changelog
https://github.com/Zulko/pianoputer/tree/master/CHANGELOG.md


CHANGELOG.md
## 2.0.2
- setup.py long_description format updated to markdown

## 2.0.1
- Fixes bug where audio files could not be loaded because mixer.init was not
  allowing the data to be modified to suit the hardware needs
- 1, 2, and 4 channel wave files now supported
- Fixes bug where keyboard files were read as ascii instead of unicode
- Fixes bug where blocked audio device events were showing up in the queues

## 2.0.0
- Add qwerty keyboard layout and defaults to using it
- Adds images for the qwery and azerty keyboards, cyan is the anchor key
- Displays the keyboard layout images when the program is running
- Quits the program if the pygame window is closed
- Adds caching of transposed audio files on the user's hard drive
- Updates bowl samples to be c6 note
- Adds a clear-cache command, invoke it with -c
- Breaking: Updates keyboard file format to end in txt and require an anchor note or key
- Changes pitch shifting to use librosa, this works for mono and stereo files
- Adds pypi installation with setup.py

## 1.0.0
- https://github.com/Zulko/pianoputer/commit/11c25b342a43af3fb367a4cda515e87a7c9a04f1


tests/test_pianoputer.py
import os
import shutil
import typing
import unittest
from unittest.mock import patch

import pygame

import pianoputer.pianoputer as pp


class PianoPuter(unittest.TestCase):
    sample_images_folder = "pianoputer/keyboards/"

    def tearDown(self):
        pygame.quit()

    def test_works_with_different_channels(self):
        audio_files = [
            "piano-c4_1channel.wav",
            "piano-c4_2channel.wav",
            "piano-c4_4channel.wav",
        ]

        def get_quit():
            return [pygame.event.Event(pygame.QUIT)]

        for audio_file in audio_files:
            with patch("pygame.event.get", side_effect=get_quit) as m_event_get:
                audio_file_path = os.path.join(os.path.dirname(__file__), audio_file)
                pp.play_pianoputer(["-w", audio_file_path])
            file_name, file_extension = os.path.splitext(audio_file)
            folder_path = os.path.join(os.path.dirname(__file__), file_name)
            shutil.rmtree(folder_path)

    def test_writes_sample_keyboards_to_images(self):
        keyboards = ["keyboards/azerty_typewriter.txt", "keyboards/qwerty_piano.txt"]
        for keyboard in keyboards:
            parser = pp.get_parser()
            args = ["-k", keyboard]
            wav_path, keyboard_path, clear_cache = pp.process_args(parser, args)
            audio_data, framerate_hz, channels = pp.get_audio_data(wav_path)
            results = pp.get_keyboard_info(keyboard_path)
            _keys, _tones, color_name_to_key, key_color, key_txt_color = results

            screen, _keyboard_graphic = pp.configure_pygame_audio_and_set_ui(
                framerate_hz,
                channels,
                keyboard_path,
                color_name_to_key,
                key_color,
                key_txt_color,
            )

            _, file_name_with_ext = os.path.split(keyboard)
            file_name, file_extension = os.path.splitext(file_name_with_ext)
            pygame.image.save(
                screen, self.sample_images_folder + "{}.jpg".format(file_name)
            )


if __name__ == "__main__":
    unittest.main()


pianoputer_stereo.py
from scipy.io import wavfile
import argparse
import numpy as np
from numpy import int16, empty, round, hanning, arange, \
    angle, fft, abs, exp, ndarray, zeros, pi
import pygame
import sys
import warnings
import timeit


def parse_arguments():
    """
    YOU CAN DEFINE THE DEFAULT SETTINGS/OPTIONS
    :return:
    """

    description = ('Use your computer keyboard as a "piano"')
    parser = argparse.ArgumentParser(description=description)
    parser.add_argument(
        '--wav', '-w',
        metavar='FILE',
        type=argparse.FileType('r'),
        # default='bowl.wav',
        default='bowlstereo.wav',
        help='WAV file (default: bowl.wav)')
    parser.add_argument(
        '--keyboard', '-k',
        metavar='FILE',
        type=argparse.FileType('r'),
        default='typewriter.kb',
        help='keyboard file (default: typewriter.kb)')
    parser.add_argument(
        '--verbose', '-v',
        action='store_true',
        help='verbose mode')

    return (parser.parse_args(), parser)


def channels(sound_object):
    """
    DETERMINE IF A SOUND OBJECT IS MONO OR STEREO
    RETURN AN INTEGER REPRESENTING THE NUMBER OF CHANNELS (1, 2, 4, 6)
    CHANNEL 4 and 6 ARE NOT IMPLEMENTED YET (NEW FEATURE IN SDL 2.0)

    :param sound_object: numpy.ndarray; Sound samples into an array
    :return: int; return the number of channels
    """
    assert isinstance(sound_object, ndarray), \
        "\nArgument sound_object is not a numpy array type, got %s " % type(sound_object)
    # CHECK MONO/STEREO
    shape = sound_object.shape
    try:
        # MULTI-CHANNELS (2, 4, 6)
        channel = shape[1]
    except IndexError:
        # MOST LIKELY TO BE MONO
        channel = 1

    if channel in (4, 6):
        # SDL 2.0 ONLY
        raise NotImplemented
    return channel

ONE_TWELVE     = 1.0 / 12.0
P2_10          = 2 << 10
P2_11          = 2 << 11
P2_12          = 2 << 12
PI2            = 2 * pi
WINDOW_SIZE    = P2_12
H              = P2_10
PHASE          = zeros(WINDOW_SIZE)
HANNING_WINDOW = hanning(WINDOW_SIZE)


# PARSE COMMAND LINE ARGUMENTS
(args, parser) = parse_arguments()

# ENABLE WARNINGS FROM SCIPY IF REQUESTED
if not args.verbose:
    warnings.simplefilter('ignore')


# Notes
#     ----- This function cannot read wav files with 24-bit data.
#
#     Common data types: [1]_
#
#     =====================  ===========  ===========  =============
#          WAV format            Min          Max       NumPy dtype
#     =====================  ===========  ===========  =============
#     32-bit floating-point  -1.0         +1.0         float32
#     32-bit PCM             -2147483648  +2147483647  int32
#     16-bit PCM             -32768       +32767       int16
#     8-bit PCM              0            255          uint8
#     =====================  ===========  ===========  =============
#
#     Note that 8-bit PCM is unsigned.
FREQ, SOUND = wavfile.read(args.wav.name)

# CHECK IF THE SOUND IS MONO OR STEREO
CHANNEL = channels(SOUND)

SOUND_LEN = len(SOUND)

TONES = range(-25, 25)
factor_list = {}
for n in TONES:
    factor_list[n] = (2 ** (n * ONE_TWELVE))


def speedx_mono(input_array, factor):
    """
    SPEEDS UP / SLOWS DOWN A SOUND, BY SOME FACTOR.
    OUTPUT ARRAY SIZES DEPENDS ON THE VARIABLE FACTOR (SEMITONES).

    :param input_array: numpy.ndarray; Sound samples into an array (1d numpy.ndarray type int16)
    :param factor: float;
    :return: Return sound samples (1d numpy.ndarray type int16)
    """
    assert isinstance(input_array, ndarray), \
        "\nArgument input_array is not a numpy array type, got %s " % type(input_array)
    assert isinstance(factor, float), \
        "\nArgument factor is not a float type, got %s " % type(factor)
    indices = round(arange(0, len(input_array), factor))
    indices = indices[indices < len(input_array)].astype(int)
    return input_array[indices]


def speedx_stereo(input_array, factor):
    """
    SPEEDS UP / SLOWS DOWN A SOUND, BY SOME FACTOR.
    OUTPUT ARRAY SIZES DEPENDS ON THE VARIABLE FACTOR (SEMITONES).

    :param input_array: numpy.ndarray; Sound samples into an array (2d array type int16)
    :param factor: float; SEMITONES
    :return: Return sound samples (2d numpy.ndarray type int16)
    """
    assert isinstance(input_array, ndarray), \
        "\nArgument input_array is not a numpy array type, got %s " % type(input_array)
    assert isinstance(factor, float), \
        "\nArgument factor is not a float type, got %s " % type(factor)
    indices = round(arange(0, len(input_array), factor))
    indices = indices[indices < len(input_array)].astype(int)
    return input_array[indices, :]

def stretch_mono(snd_array, factor):
    """
    STRETCHES/SHORTENS A SOUND, BY SOME FACTOR.

    SOUND STRETCHING CAN BE DONE USING THE CLASSICAL PHASE VOCODER METHOD.
    YOU FIRST BREAK THE SOUND INTO OVERLAPPING BITS, AND YOU REARRANGE THESE BITS
    SO THAT THEY WILL OVERLAP EVEN MORE (IF YOU WANT TO SHORTEN THE SOUND)
    OR LESS (IF YOU WANT TO STRETCH THE SOUND)

    * COMPATIBLE ONLY WITH MONO SOUND

    :param snd_array: numpy.ndarray; Sound samples into an array (1d array type int16)
    :param factor: float; SEMITONES
    :return: return 1d numpy.ndarray type int16
    """
    assert isinstance(snd_array, ndarray), \
        "\nArgument input_array is not a numpy array type, got %s " % type(snd_array)
    assert isinstance(factor, float), \
        "\nArgument factor is not a float type, got %s " % type(factor)

    """ Stretches/shortens a SOUND, by some factor. """
    phase = PHASE.copy()
    result = zeros(int(SOUND_LEN / factor + WINDOW_SIZE))

    for i in arange(0, SOUND_LEN - (WINDOW_SIZE + H), H*factor):
        i = int(i)
        # TWO POTENTIALLY OVERLAPPING SUBARRAY
        a1 = snd_array[i: i + WINDOW_SIZE]
        a2 = snd_array[i + H: i + WINDOW_SIZE + H]

        # The spectra of these arrays
        s1 = fft.fft(HANNING_WINDOW * a1)
        s2 = fft.fft(HANNING_WINDOW * a2)

        # REPHASE ALL FREQUENCIES
        phase = (phase + angle(s2/s1)) % PI2

        a2_rephased = fft.ifft(abs(s2)*exp(1j*phase))
        i2 = int(i/factor)
        result[i2: i2 + WINDOW_SIZE] += HANNING_WINDOW*a2_rephased.real

    # NORMALIZE (16BIT)
    result = (P2_11 * result/result.max())

    return result.astype('int16')


def stretch_stereo(snd_array, factor):
    """
    STRETCHES/SHORTENS A SOUND, BY SOME FACTOR.

    SOUND STRETCHING CAN BE DONE USING THE CLASSICAL PHASE VOCODER METHOD.
    YOU FIRST BREAK THE SOUND INTO OVERLAPPING BITS, AND YOU REARRANGE THESE BITS
    SO THAT THEY WILL OVERLAP EVEN MORE (IF YOU WANT TO SHORTEN THE SOUND)
    OR LESS (IF YOU WANT TO STRETCH THE SOUND)

    * COMPATIBLE ONLY FOR STEREO SOUNDS:
    IN STEREOPHONIC SOUND MORE CHANNELS ARE USED (TYPICALLY TWO).
    YOU CAN USE TWO DIFFERENT CHANNELS AND MAKE ONE FEED ONE SPEAKER
    AND THE SECOND CHANNEL FEED A SECOND SPEAKER (WHICH IS THE MOST COMMON STEREO SETUP).
    THIS IS USED TO CREATE DIRECTIONALITY, PERSPECTIVE, SPACE.
    IF YOU’VE EVER LOOKED AT THE WAVEFORM OF A STEREO AUDIO FILE WITHIN A DIGITAL AUDIO WORKSTATION (DAW),
    YOU’VE LIKELY NOTICED THAT THERE ARE TWO WAVEFORMS APART OF THE FILE.
    EACH WAVEFORM REPRESENTS A SINGLE CHANNEL OF AUDIO.

    :param snd_array: numpy.ndarray; Sound samples into an array (2d array type int16)
    :param factor: float; SEMITONES
    :return: return a 2d numpy.ndarray type int16
    """

    assert isinstance(snd_array, ndarray), \
        "\nArgument input_array is not a numpy array type, got %s " % type(snd_array)
    # CHECK CHANNELS
    assert snd_array.shape[1] == 2, \
        "\nArgument input_array is not a valid SOUND sample data," \
        " expecting array shape (w, 2) got (%s, %s)" % snd_array.shape
    assert isinstance(factor, float), \
        "\nArgument factor is not a float type, got %s " % type(factor)

    result = zeros((int(SOUND_LEN / factor + WINDOW_SIZE), 2))
    phase1 = PHASE.copy()
    phase2 = PHASE.copy()

    for i in arange(0, SOUND_LEN - (WINDOW_SIZE + H), H * factor):
        i = int(i)
        # EXTRACT CHANNEL 1 & 2 SECTION (I: I + WINDOW_SIZE)
        a1_ch1 = snd_array[i: i + WINDOW_SIZE, 0]
        a1_ch2 = snd_array[i: i + WINDOW_SIZE, 1]
        # COMPUTE THE ONE-DIMENSIONAL DISCRETE FOURIER
        # TRANSFORM FOR CHANNEL 1 & 2 SECTION (I: I + WINDOW_SIZE)
        s1  = fft.fft(HANNING_WINDOW * a1_ch1)
        s11 = fft.fft(HANNING_WINDOW * a1_ch2)

        # EXTRACT CHANNEL 1 & 2 SECTION (i + H: i + WINDOW_SIZE + H)
        a2_ch1 = snd_array[i + H: i + WINDOW_SIZE + H, 0]
        a2_ch2 = snd_array[i + H: i + WINDOW_SIZE + H, 1]
        # COMPUTE THE ONE-DIMENSIONAL DISCRETE FOURIER
        # TRANSFORM FOR CHANNEL 1 & 2 SECTION (I + H: I + WINDOW_SIZE + H)
        s2  = fft.fft(HANNING_WINDOW * a2_ch1)
        s22 = fft.fft(HANNING_WINDOW * a2_ch2)

        phase1 = (phase1 + angle(s2 / s1)) % PI2
        phase2 = (phase2 + angle(s22 / s11)) % PI2

        a2_rephased  = fft.ifft(abs(s2) * exp(1j * phase1))
        a22_rephased = fft.ifft(abs(s22) * exp(1j * phase2))

        i2 = int(i / factor)

        result[i2: i2 + WINDOW_SIZE, 0] += HANNING_WINDOW * a2_rephased.real
        result[i2: i2 + WINDOW_SIZE, 1] += HANNING_WINDOW * a22_rephased.real

    # NORMALIZE (16BIT)
    result = (P2_12 * result / result.max())

    return result.astype(int16)


def pitchshift(snd_array, n):
    """
    PITCH-SHIFTING IS EASY ONCE YOU HAVE SOUND STRETCHING.
    IF YOU WANT A HIGHER PITCH, YOU FIRST STRETCH THE SOUND WHILE CONSERVING THE PITCH,
    THEN YOU SPEED UP THE RESULT, SUCH THAT THE FINAL SOUND HAS THE SAME DURATION AS
    THE INITIAL ONE, BUT A HIGHER PITCH DUE TO THE SPEED CHANGE.

    :param snd_array: numpy.ndarray; Sound samples into an array (2d array type int16)
    :param n: float; SEMITONES
    :return: return a numpy.ndarray that can be converted into a pygame.sound object
    """

    """ CHANGES THE PITCH OF A SOUND BY ``N`` SEMITONES. """
    assert isinstance(snd_array, ndarray), \
        "\nArgument input_array is not a numpy array type, got %s " % type(snd_array)
    assert isinstance(n, int), \
        "\nArgument n is not a int type, got %s " % type(n)

    # SELECT THE CORRECT ALGORITHM ACCORDING TO THE NUMBER OF TRACK(S)
    if CHANNEL == 2:
        stretched = stretch_stereo(snd_array, 1.0/factor_list[n])
        return speedx_stereo(stretched[WINDOW_SIZE:], factor_list[n])
    else:
        stretched = stretch_mono(snd_array, 1.0/factor_list[n])
        return speedx_mono(stretched[WINDOW_SIZE:], factor_list[n])


def main():

    sys.stdout.write('Transponding SOUND file... ')
    sys.stdout.flush()
    transposed_sounds = [pitchshift(SOUND, n) for n in TONES]
    print('DONE')

    # So flexible ;)
    # The channels argument is used to specify whether to use mono or stereo.
    # 1 for mono and 2 for stereo.
    # New in pygame 2: The number of channels can also be 4 or 6.
    # Some platforms require the pygame.mixer pygame module for loading and playing
    # sounds module to be initialized after the display modules have initialized.
    # The top level pygame.init() takes care of this automatically, but cannot pass
    # any arguments to the mixer init. To solve this, mixer has a function pygame.mixer.pre_init()
    # to set the proper defaults before the toplevel init is used.

    # The pygame.mixer.init() function takes several optional arguments to control
    # the playback rate and sample size. Pygame will default to reasonable values,
    # but pygame cannot perform Sound ressampling, so the mixer should be initialized
    # to match the values of your audio resources (This is why we are
    # using FREQ, SOUND = wavfile.read(args.wav.name) to extract the sample rate (FREQ)
    pygame.mixer.init(frequency=FREQ, size=-16, channels=CHANNEL, buffer=2048)

    pygame.display.set_mode((250, 250))

    keys = args.keyboard.read().split('\n')

    # CREATE A NEW PLAYABLE SOUND OBJECT FROM AN ARRAY.
    # THE MIXER MODULE MUST BE INITIALIZED AND THE ARRAY FORMAT MUST BE SIMILAR TO THE MIXER AUDIO FORMAT.
    sounds = map(pygame.sndarray.make_sound, transposed_sounds)

    key_sound = dict(zip(keys, sounds))
    is_playing = {k: False for k in keys}

    # todo if the window lost focus cant hear the SOUND
    while True:

        event = pygame.event.wait()

        if event.type in (pygame.KEYDOWN, pygame.KEYUP):
            key = pygame.key.name(event.key)

        if event.type == pygame.KEYDOWN:
            if (key in key_sound.keys()) and (not is_playing[key]):
                key_sound[key].play(fade_ms=50)
                is_playing[key] = True

            elif event.key == pygame.K_ESCAPE:
                pygame.quit()
                raise KeyboardInterrupt

        elif event.type == pygame.KEYUP and key in key_sound.keys():
            # Stops with 50ms fadeout
            key_sound[key].fadeout(50)
            is_playing[key] = False


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('Goodbye')


pianoputer/pianoputer.py
#!/usr/bin/env python

import argparse
import codecs
import os
import re
import shutil
import warnings
from collections import defaultdict
from pathlib import Path
from typing import Dict, Generator, List, Optional, Tuple

import keyboardlayout as kl
import keyboardlayout.pygame as klp
import librosa
import numpy
import pygame
import soundfile

ANCHOR_INDICATOR = " anchor"
ANCHOR_NOTE_REGEX = re.compile(r"\s[abcdefg]$")
DESCRIPTION = 'Use your computer keyboard as a "piano"'
DESCRIPTOR_32BIT = "FLOAT"
BITS_32BIT = 32
AUDIO_ALLOWED_CHANGES_HARDWARE_DETERMINED = 0
SOUND_FADE_MILLISECONDS = 50
CYAN = (0, 255, 255, 255)
BLACK = (0, 0, 0, 255)
WHITE = (255, 255, 255, 255)

AUDIO_ASSET_PREFIX = "audio_files/"
KEYBOARD_ASSET_PREFIX = "keyboards/"
CURRENT_WORKING_DIR = Path(__file__).parent.absolute()
ALLOWED_EVENTS = {pygame.KEYDOWN, pygame.KEYUP, pygame.QUIT}


def get_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=DESCRIPTION)
    default_wav_file = "audio_files/piano_c4.wav"
    parser.add_argument(
        "--wav",
        "-w",
        metavar="FILE",
        type=str,
        default=default_wav_file,
        help="WAV file (default: {})".format(default_wav_file),
    )
    default_keyboard_file = "keyboards/qwerty_piano.txt"
    parser.add_argument(
        "--keyboard",
        "-k",
        metavar="FILE",
        type=str,
        default=default_keyboard_file,
        help="keyboard file (default: {})".format(default_keyboard_file),
    )
    parser.add_argument(
        "--clear-cache",
        "-c",
        default=False,
        action="store_true",
        help="deletes stored transposed audio files and recalculates them",
    )
    parser.add_argument("--verbose", "-v", action="store_true", help="verbose mode")

    return parser


def get_or_create_key_sounds(
    wav_path: str,
    sample_rate_hz: int,
    channels: int,
    tones: List[int],
    clear_cache: bool,
    keys: List[str],
) -> Generator[pygame.mixer.Sound, None, None]:
    sounds = []
    y, sr = librosa.load(wav_path, sr=sample_rate_hz, mono=channels == 1)
    file_name = os.path.splitext(os.path.basename(wav_path))[0]
    folder_containing_wav = Path(wav_path).parent.absolute()
    cache_folder_path = Path(folder_containing_wav, file_name)
    if clear_cache and cache_folder_path.exists():
        shutil.rmtree(cache_folder_path)
    if not cache_folder_path.exists():
        print("Generating samples for each key")
        os.mkdir(cache_folder_path)
    for i, tone in enumerate(tones):
        cached_path = Path(cache_folder_path, "{}.wav".format(tone))
        if Path(cached_path).exists():
            print("Loading note {} out of {} for {}".format(i + 1, len(tones), keys[i]))
            sound, sr = librosa.load(cached_path, sr=sample_rate_hz, mono=channels == 1)
            if channels > 1:
                # the shape must be [length, 2]
                sound = numpy.transpose(sound)
        else:
            print(
                "Transposing note {} out of {} for {}".format(
                    i + 1, len(tones), keys[i]
                )
            )
            if channels == 1:
                sound = librosa.effects.pitch_shift(y, sr, n_steps=tone)
            else:
                new_channels = [
                    librosa.effects.pitch_shift(y[i], sr, n_steps=tone)
                    for i in range(channels)
                ]
                sound = numpy.ascontiguousarray(numpy.vstack(new_channels).T)
            soundfile.write(cached_path, sound, sample_rate_hz, DESCRIPTOR_32BIT)
        sounds.append(sound)
    sounds = map(pygame.sndarray.make_sound, sounds)
    return sounds


BLACK_INDICES_C_SCALE = [1, 3, 6, 8, 10]
LETTER_KEYS_TO_INDEX = {"c": 0, "d": 2, "e": 4, "f": 5, "g": 7, "a": 9, "b": 11}


def __get_black_key_indices(key_name: str) -> set:
    letter_key_index = LETTER_KEYS_TO_INDEX[key_name]
    black_key_indices = set()
    for ind in BLACK_INDICES_C_SCALE:
        new_index = ind - letter_key_index
        if new_index < 0:
            new_index += 12
        black_key_indices.add(new_index)
    return black_key_indices


def get_keyboard_info(keyboard_file: str):
    with codecs.open(keyboard_file, encoding="utf-8") as k_file:
        lines = k_file.readlines()
    keys = []
    anchor_index = -1
    black_key_indices = set()
    for i, line in enumerate(lines):
        line = line.strip()
        if not line:
            continue
        match = ANCHOR_NOTE_REGEX.search(line)
        if match:
            anchor_index = i
            black_key_indices = __get_black_key_indices(line[-1])
            key = kl.Key(line[: match.start(0)])
        elif line.endswith(ANCHOR_INDICATOR):
            anchor_index = i
            key = kl.Key(line[: -len(ANCHOR_INDICATOR)])
        else:
            key = kl.Key(line)
        keys.append(key)
    if anchor_index == -1:
        raise ValueError(
            "Invalid keyboard file, one key must have an anchor note or the "
            "word anchor written next to it.\n"
            "For example 'm c OR m anchor'.\n"
            "That tells the program that the wav file will be used for key m, "
            "and all other keys will be pitch shifted higher or lower from "
            "that anchor. If an anchor note is used then keys are colored black "
            "and white like a piano. If the word anchor is used, then the "
            "highest key is white, and keys get darker as they descend in pitch."
        )
    tones = [i - anchor_index for i in range(len(keys))]
    color_to_key = defaultdict(list)
    if black_key_indices:
        key_color = (120, 120, 120, 255)
        key_txt_color = (50, 50, 50, 255)
    else:
        key_color = (65, 65, 65, 255)
        key_txt_color = (0, 0, 0, 255)
    for index, key in enumerate(keys):
        if index == anchor_index:
            color_to_key[CYAN].append(key)
            continue
        if black_key_indices:
            used_index = (index - anchor_index) % 12
            if used_index in black_key_indices:
                color_to_key[BLACK].append(key)
                continue
            color_to_key[WHITE].append(key)
            continue
        # anchor mode, keys go up in half steps and we do not color black keys
        # instead we color from grey low to white high
        rgb_val = 255 - (len(keys) - 1 - index) * 3
        color = (rgb_val, rgb_val, rgb_val, 255)
        color_to_key[color].append(key)

    return keys, tones, color_to_key, key_color, key_txt_color


def configure_pygame_audio_and_set_ui(
    framerate_hz: int,
    channels: int,
    keyboard_arg: str,
    color_to_key: Dict[str, List[kl.Key]],
    key_color: Tuple[int, int, int, int],
    key_txt_color: Tuple[int, int, int, int],
) -> Tuple[pygame.Surface, klp.KeyboardLayout]:
    # ui
    pygame.display.init()
    pygame.display.set_caption("pianoputer")

    # block events that we don't want, this must be after display.init
    pygame.event.set_blocked(None)
    pygame.event.set_allowed(list(ALLOWED_EVENTS))

    # fonts
    pygame.font.init()

    # audio
    pygame.mixer.init(
        framerate_hz,
        BITS_32BIT,
        channels,
        allowedchanges=AUDIO_ALLOWED_CHANGES_HARDWARE_DETERMINED,
    )

    screen_width = 50
    screen_height = 50
    if "qwerty" in keyboard_arg:
        layout_name = kl.LayoutName.QWERTY
    elif "azerty" in keyboard_arg:
        layout_name = kl.LayoutName.AZERTY_LAPTOP
    else:
        ValueError("keyboard must have qwerty or azerty in its name")
    margin = 4
    key_size = 60
    overrides = {}
    for color_value, keys in color_to_key.items():
        override_color = color = pygame.Color(color_value)
        inverted_color = list(~override_color)
        other_val = 255
        if (
            abs(color_value[0] - inverted_color[0]) > abs(color_value[0] - other_val)
        ) or color_value == CYAN:
            override_txt_color = pygame.Color(inverted_color)
        else:
            # biases grey override keys to use white as txt_color
            override_txt_color = pygame.Color([other_val] * 3 + [255])
        override_key_info = kl.KeyInfo(
            margin=margin,
            color=override_color,
            txt_color=override_txt_color,
            txt_font=pygame.font.SysFont("Arial", key_size // 4),
            txt_padding=(key_size // 10, key_size // 10),
        )
        for key in keys:
            overrides[key.value] = override_key_info

    key_txt_color = pygame.Color(key_txt_color)
    keyboard_info = kl.KeyboardInfo(position=(0, 0), padding=2, color=key_txt_color)
    key_info = kl.KeyInfo(
        margin=margin,
        color=pygame.Color(key_color),
        txt_color=pygame.Color(key_txt_color),
        txt_font=pygame.font.SysFont("Arial", key_size // 4),
        txt_padding=(key_size // 6, key_size // 10),
    )
    letter_key_size = (key_size, key_size)  # width, height
    keyboard = klp.KeyboardLayout(
        layout_name, keyboard_info, letter_key_size, key_info, overrides
    )
    screen_width = keyboard.rect.width
    screen_height = keyboard.rect.height

    screen = pygame.display.set_mode((screen_width, screen_height))
    screen.fill(pygame.Color("black"))
    if keyboard:
        keyboard.draw(screen)
    pygame.display.update()
    return screen, keyboard


def play_until_user_exits(
    keys: List[kl.Key],
    key_sounds: List[pygame.mixer.Sound],
    keyboard: klp.KeyboardLayout,
):
    sound_by_key = dict(zip(keys, key_sounds))
    playing = True

    while playing:
        for event in pygame.event.get():

            if event.type == pygame.QUIT:
                playing = False
                break
            elif event.key == pygame.K_ESCAPE:
                playing = False
                break

            key = keyboard.get_key(event)
            if key is None:
                continue
            try:
                sound = sound_by_key[key]
            except KeyError:
                continue

            if event.type == pygame.KEYDOWN:
                sound.stop()
                sound.play(fade_ms=SOUND_FADE_MILLISECONDS)
            elif event.type == pygame.KEYUP:
                sound.fadeout(SOUND_FADE_MILLISECONDS)

    pygame.quit()
    print("Goodbye")


def get_audio_data(wav_path: str) -> Tuple:
    audio_data, framerate_hz = soundfile.read(wav_path)
    array_shape = audio_data.shape
    if len(array_shape) == 1:
        channels = 1
    else:
        channels = array_shape[1]
    return audio_data, framerate_hz, channels


def process_args(parser: argparse.ArgumentParser, args: Optional[List]) -> Tuple:
    if args:
        args = parser.parse_args(args)
    else:
        args = parser.parse_args()

    # Enable warnings from scipy if requested
    if not args.verbose:
        warnings.simplefilter("ignore")

    wav_path = args.wav
    if wav_path.startswith(AUDIO_ASSET_PREFIX):
        wav_path = os.path.join(CURRENT_WORKING_DIR, wav_path)

    keyboard_path = args.keyboard
    if keyboard_path.startswith(KEYBOARD_ASSET_PREFIX):
        keyboard_path = os.path.join(CURRENT_WORKING_DIR, keyboard_path)
    return wav_path, keyboard_path, args.clear_cache


def play_pianoputer(args: Optional[List[str]] = None):
    parser = get_parser()
    wav_path, keyboard_path, clear_cache = process_args(parser, args)
    audio_data, framerate_hz, channels = get_audio_data(wav_path)
    results = get_keyboard_info(keyboard_path)
    keys, tones, color_to_key, key_color, key_txt_color = results
    key_sounds = get_or_create_key_sounds(
        wav_path, framerate_hz, channels, tones, clear_cache, keys
    )

    _screen, keyboard = configure_pygame_audio_and_set_ui(
        framerate_hz, channels, keyboard_path, color_to_key, key_color, key_txt_color
    )
    print(
        "Ready for you to play!\n"
        "Press the keys on your keyboard. "
        "To exit presss ESC or close the pygame window"
    )
    play_until_user_exits(keys, key_sounds, keyboard)


if __name__ == "__main__":
    play_pianoputer()
